hal/hal_remote/include/hal_remote/hal_client.h

```diff
@@ -0,0 +1,115 @@
+// Copyright (C) Codeplay Software Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License") with LLVM
+// Exceptions; you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://github.com/codeplaysoftware/oneapi-construction-kit/blob/main/LICENSE.txt
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef _HAL_CLIENT_H
+#define _HAL_CLIENT_H
+
+#include <hal.h>
+#include <hal_remote/hal_binary_decoder.h>
+#include <hal_remote/hal_binary_encoder.h>
+#include <hal_remote/hal_device_client.h>
+
+#include <mutex>
+
+namespace hal {
+
+/// @brief Base class for hal_client, is a type of `hal::hal_t`, which can be
+/// used to communicate with a remote server and will create `hal_device_client`
+/// when requested via requests to the server.
+/// @note It currently has the limitation of only supporting a single device.
+class hal_client : public hal::hal_t {
+ protected:
+  std::mutex lock;
+  bool made_connection = false;
+  hal::hal_info_t hal_info;
+  hal::hal_device_info_t hal_device_info;
+  uint16_t required_port;
+
+ public:
+  hal_client() {}
+  virtual bool make_connection() = 0;
+  virtual hal::hal_transmitter &get_transmitter() = 0;
+  // request the creation of a new hal
+  hal::hal_device_t *device_create(uint32_t index) override {
+    const std::lock_guard<std::mutex> locker(lock);
+    if (index > 0) {
+      return nullptr;
+    }
+
+    if (!made_connection) {
+      if (!make_connection()) {
+        return nullptr;
+      }
+      made_connection = true;
+    }
+
+    hal::hal_binary_encoder encoder(index);
+    hal::hal_binary_decoder decoder;
+    hal::hal_binary_encoder::COMMAND command;
+    encoder.encode_device_create();
+    get_transmitter().send(encoder.data(), encoder.size(), true);
+    bool ok = get_transmitter().receive(
+        (uint8_t *)&command, sizeof(hal::hal_binary_encoder::COMMAND));
```

Comment:
This is wrong if server and client disagree on endianness. We only have little endian targets at the moment, can we add an assert for that?

Reply:
I agree it depends on matching endianness. I'll have to have a think about best to check the endianness with an assert (in theory it's in the info).